test/channels.jl

```diff
@@ -578,3 +579,39 @@ let c = Channel(3)
     close(c)
     @test repr(MIME("text/plain"), c) == "Channel{Any}(3) (closed)"
 end
+
+# PR #41833: data races in Channel
+@testset "n_avail(::Channel)" begin
+    # Buffered: n_avail() = buffer length + number of waiting tasks
+    let c = Channel(2)
+        @test n_avail(c) == 0;   put!(c, 0)
+        @test n_avail(c) == 1;   put!(c, 0)
+        @test n_avail(c) == 2;   yield(@task put!(c, 0))
+        @test n_avail(c) == 3;   yield(@task put!(c, 0))
+        @test n_avail(c) == 4
+        # Test n_avail(c) after interrupting a task waiting on the channel
+                                t = @task put!(c, 0)
+                                yield(t)
+        @test n_avail(c) == 5
+                                @async Base.throwto(t, ErrorException("Exit put!"))
+                                try wait(t) catch end
+        @test n_avail(c) == 4
+                                close(c); yield()
+        @test n_avail(c) == 2    # Already-buffered items remain
```

Comment:
I think it's safer/more robust to explicitly `wait` on the tasks rather than `close(c); yield()`? I think the scheduler is allowed to not process all tasks upon a single `yield`?

```suggestion
        @test n_avail(c) == 2;   t1 = @task put!(c, 0); yield(t1)
        @test n_avail(c) == 3;   t2 = @task put!(c, 0); yield(t2)
        @test n_avail(c) == 4
        # Test n_avail(c) after interrupting a task waiting on the channel
                                t3 = @task put!(c, 0)
                                yield(t3)
        @test n_avail(c) == 5
                                @async Base.throwto(t3, ErrorException("Exit put!"))
                                try wait(t3) catch end
        @test n_avail(c) == 4
                                close(c)
                                try wait(t1) catch end
                                try wait(t2) catch end
        @test n_avail(c) == 2    # Already-buffered items remain
```

Reply:
I don't know about the current rules for `yield()` but that does seem like an implementation choice.

An explicit `wait()` should always be better.